File: packages/analytics/__tests__/sinks.test.ts
import { describe, it, expect, vi } from 'vitest'
import { createConsoleSink } from '../src/console-sink.js'
import { createMockSink } from '../src/mock-sink.js'
import { createNoopAnalytics } from '../src/noop.js'
import { SCHEMA_VERSION } from '../src/types.js'
import type { AnalyticsEvent } from '../src/types.js'

function ev(event: string): AnalyticsEvent {
  return {
    type: 'track',
    event,
    messageId: 'm',
    anonymousId: 'a',
    sessionId: 's',
    properties: {},
    context: {
      app: 'app',
      env: 'test',
      library: { name: '@refraction-ui/analytics', version: '0.1.0' },
    },
    timestamp: new Date().toISOString(),
    schemaVersion: SCHEMA_VERSION,
  }
}

describe('createConsoleSink', () => {
  it('groups and logs each canonical envelope', () => {
    const logger = {
      log: vi.fn(),
      groupCollapsed: vi.fn(),
      groupEnd: vi.fn(),
    }
    const sink = createConsoleSink({ logger })
    sink.deliver([ev('A'), ev('B')], { unload: false })
    expect(logger.groupCollapsed).toHaveBeenCalledTimes(2)
    expect(logger.log).toHaveBeenCalledTimes(2)
    expect(logger.groupCollapsed).toHaveBeenCalledWith('[analytics] track A')
  })

  it('falls back to plain log when grouping is unavailable', () => {
    const log = vi.fn()
    const sink = createConsoleSink({
      logger: { log } as never,
    })
    sink.deliver([ev('Solo')], { unload: false })
    expect(log).toHaveBeenCalledWith(
      '[analytics] track Solo',
      expect.objectContaining({ event: 'Solo' }),
    )
  })

  it('passes through configured consent categories', () => {
    const sink = createConsoleSink({ consentCategories: ['debug'] })
    expect(sink.consentCategories).toEqual(['debug'])
  })
})

describe('createMockSink', () => {
  it('captures init/deliver/flush/shutdown for assertions', async () => {
    const sink = createMockSink({ name: 's1', consentCategories: ['x'] })
    expect(sink.name).toBe('s1')
    expect(sink.consentCategories).toEqual(['x'])

    sink.init?.({ app: 'a', env: 'e' })
    sink.deliver([ev('A')], { unload: false })
    sink.deliver([ev('B'), ev('C')], { unload: true })
    await sink.flush?.()
    await sink.shutdown?.()

    expect(sink.initCalls).toHaveLength(1)
    expect(sink.deliveries).toHaveLength(2)
    expect(sink.deliveries[1].ctx.unload).toBe(true)
    expect(sink.events.map((e) => e.event)).toEqual(['A', 'B', 'C'])
    expect(sink.flushCalls).toBe(1)
    expect(sink.shutdownCalls).toBe(1)
  })
})

describe('createNoopAnalytics', () => {
  it('exposes the full Analytics surface, all inert', () => {
    const a = createNoopAnalytics()
    expect(a.enabled).toBe(false)
    expect(a.sinks).toEqual([])
    expect(a.userId()).toBeUndefined()
    expect(a.anonymousId()).toMatch(/^0{8}-0{4}-4/)
    expect(typeof a.session.id()).toBe('string')
    expect(a.consent.granted()).toEqual([])
    expect(a.consent.isGranted('analytics')).toBe(false)
    expect(a.with({})).toBe(a)
    expect(() => {
      a.track('x')
      a.identify('u')
      a.page()
      a.screen()
      a.group('g')
      a.alias('u2')
      a.session.start()
      a.session.end()
      a.session.set({})
      a.consent.grant('a')
      a.consent.revoke('a')
      a.addSink(createMockSink())
      a.removeSink('mock')
      a.reset()
    }).not.toThrow()
  })

  it('flush resolves without doing anything', async () => {
    const a = createNoopAnalytics()
    await expect(a.flush()).resolves.toBeUndefined()
  })
})

File: packages/flutter/test/analytics/sinks_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:refraction_ui/refraction_ui.dart'
    hide createConsoleSink, createMockSink, MockSinkExtended;
// Analytics' own console sink — the package barrel hides it (telemetry owns
// the flat-barrel createConsoleSink), so import it directly here.
import 'package:refraction_ui/src/analytics/console_sink.dart';
// Test-only mock sink — intentionally not re-exported anywhere; would collide
// with telemetry's flat-barrel createMockSink / MockSinkExtended.
import 'package:refraction_ui/src/analytics/mock_sink.dart';

AnalyticsEvent _ev(String event) => AnalyticsEvent(
  type: AnalyticsEventType.track,
  event: event,
  messageId: 'm',
  anonymousId: 'a',
  sessionId: 's',
  properties: const {},
  context: const AnalyticsContext(
    app: 'app',
    env: 'test',
    library: AnalyticsLibrary(
      name: '@refraction-ui/analytics',
      version: '0.1.0',
    ),
  ),
  timestamp: DateTime.now().toUtc().toIso8601String(),
  schemaVersion: kSchemaVersion,
);

void main() {
  group('ConsoleSink', () {
    test('logs a labelled line per canonical envelope', () {
      final lines = <String>[];
      final sink = createConsoleSink(logger: lines.add);
      sink.deliver([
        _ev('A'),
        _ev('B'),
      ], const SinkDeliverContext(unload: false));
      expect(lines, hasLength(2));
      expect(lines[0], startsWith('[analytics] track A'));
      expect(lines[1], startsWith('[analytics] track B'));
    });

    test('passes through configured consent categories', () {
      final sink = createConsoleSink(consentCategories: ['debug']);
      expect(sink.consentCategories, ['debug']);
    });
  });

  group('MockSink', () {
    test('captures init/deliver/flush/shutdown for assertions', () {
      final sink = createMockSink(name: 's1', consentCategories: ['x']);
      expect(sink.name, 's1');
      expect(sink.consentCategories, ['x']);

      sink.init(const SinkInitContext(app: 'a', env: 'e'));
      sink.deliver([_ev('A')], const SinkDeliverContext(unload: false));
      sink.deliver([
        _ev('B'),
        _ev('C'),
      ], const SinkDeliverContext(unload: true));
      sink.flush();
      sink.shutdown();

      expect(sink.initCalls, hasLength(1));
      expect(sink.deliveries, hasLength(2));
      expect(sink.deliveries[1].ctx.unload, isTrue);
      expect(sink.events.map((e) => e.event), ['A', 'B', 'C']);
      expect(sink.flushCalls, 1);
      expect(sink.shutdownCalls, 1);
    });
  });

  group('createNoopAnalytics', () {
    test('exposes the full Analytics surface, all inert', () async {
      final a = createNoopAnalytics();
      expect(a.enabled, isFalse);
      expect(a.sinks, isEmpty);
      expect(a.userId(), isNull);
      expect(a.anonymousId(), startsWith('00000000-0000-4'));
      expect(a.session.id(), isA<String>());
      expect(a.consent.granted(), isEmpty);
      expect(a.consent.isGranted('analytics'), isFalse);
      expect(a.withContext(const {}), same(a));
      expect(() {
        a.track('x');
        a.identify('u');
        a.page();
        a.screen();
        a.group('g');
        a.alias('u2');
        a.session.start();
        a.session.end();
        a.session.set(const {});
        a.consent.grant(['a']);
        a.consent.revoke(['a']);
        a.addSink(createMockSink());
        a.removeSink('mock');
        a.reset();
      }, returnsNormally);
      await a.flush();
    });
  });
}
